ste/destReauthPolicy.go

```diff
@@ -0,0 +1,114 @@
+package ste
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/bloberror"
+	"github.com/Azure/azure-storage-azcopy/v10/common"
+	"net/http"
+	"sync"
+	"time"
+)
+
+/*
+RESPONSE Status: 401 Server failed to authenticate the request. Please refer to the information in the www-authenticate header.
+Www-Authenticate: Bearer authorization_uri=https://login.microsoftonline.com/72f988bf-86f1-41af-91ab-2d7cd011db47/oauth2/authorize resource_id=https://storage.azure.com
+X-Ms-Error-Code: InvalidAuthenticationInfo
+<?xml version="1.0" encoding="utf-8"?> <Error><Code>InvalidAuthenticationInfo</Code><Message>Server failed to authenticate the request. Please refer to the information in the www-authenticate header. </Message><AuthenticationErrorDetail>Lifetime validation failed. The token is expired.</AuthenticationErrorDetail>
+*/
+
+type destReauthPolicy struct {
+	cred *common.ScopedAuthenticator
+}
+
+var reauthLock *sync.Cond = sync.NewCond(&sync.Mutex{})
+
+func NewDestReauthPolicy(cred *common.ScopedAuthenticator) policy.Policy {
+	return &destReauthPolicy{cred}
+}
+
+func (d *destReauthPolicy) Do(req *policy.Request) (*http.Response, error) {
+retry:
+	ctx := req.Raw().Context()
+	debugCtx := context.WithValue(ctx, "destAuthDebug", true)
+
+	clone := req.Clone(ctx)
+	resp, err := clone.Next() // Initially attempt the request.
+
+	if err != nil || resp.StatusCode != http.StatusOK { // But, if we get back an error...
+		var authReq *azidentity.AuthenticationRequiredError
+		var respErr = &azcore.ResponseError{}
+
+		reauth := false
+
+		switch { // Is it an error we can resolve by re-authing?
+		case errors.As(err, &authReq):
+			reauth = true
+			debugCtx = context.WithValue(debugCtx, "reauthSrc", "AuthenticationRequiredError")
+		case resp.StatusCode == http.StatusUnauthorized:
+			errors.As(runtime.NewResponseError(resp), &respErr)
+			reauth = err == nil &&
+				bloberror.HasCode(respErr, bloberror.InvalidAuthenticationInfo) &&
+				len(respErr.RawResponse.Header.Values("WWW-Authenticate")) != 0
+			if reauth {
+				debugCtx = context.WithValue(debugCtx, "reauthSrc", "InvalidAuthenticationInfo")
+			}
+		}
+
+		if reauth { // If it is, pull the lock if we can, reauth
+			m := reauthLock.L.(*sync.Mutex)
+
+			if m.TryLock() { // Fetch the lock and try until we get auth.
+				for {
+					if ctx.Value("noPrompt") == nil {
```

Comment:
Is this purely for test?

Reply:
Yep. We want to test the policy, and not get caught up waiting for a response that can never come.